pkg/controller/cronjob/utils.go

```diff
@@ -106,9 +106,25 @@ func mostRecentScheduleTime(cj *batchv1.CronJob, now time.Time, schedule cron.Sc
 	if timeBetweenTwoSchedules < 1 {
 		return earliestTime, nil, 0, fmt.Errorf("time difference between two schedules is less than 1 second")
 	}
+	// this logic used for calculating number of missed schedules does a rough
+	// approximation, by calculating a diff between two schedules (t1 and t2),
+	// and counting how many of these will fit in between last schedule and now
 	timeElapsed := int64(now.Sub(t1).Seconds())
 	numberOfMissedSchedules := (timeElapsed / timeBetweenTwoSchedules) + 1
-	mostRecentTime := time.Unix(t1.Unix()+((numberOfMissedSchedules-1)*timeBetweenTwoSchedules), 0).UTC()
+
+	var mostRecentTime time.Time
+	// to get the most recent time accurate for regular schedules and the ones
+	// specified with @every form we first need to calculate potential earliest
+	// time by multipliying number of missed schedules by its interval, this
+	// is critical to ensure @every starts at the correct time, this explains
+	// the numberOfMissedSchedules-1, the additional -1 serves there to go back
+	// in time one more time unit back and let the cron library calculate
+	// a proper schedule, for case where the schedule is not consistent,
+	// for example something like  30 6-16/4 * * 1-5
+	potentialEarliest := t1.Add(time.Duration((numberOfMissedSchedules-1-1)*timeBetweenTwoSchedules) * time.Second)
```

Comment:
Upon second thought, this will not work in few corner scenarios.If the schedule is such that if the difference between now and the actual most recent time is greater than twice the timeBetweenTwoSchedules, then mostRecentTime would be returned as empty. An example test case which would fail would be 
```
{
			name: "complex schedule",
			cj: &batchv1.CronJob{
				ObjectMeta: metav1.ObjectMeta{
					CreationTimestamp: metav1TopOfTheHour,
				},
				Spec: batchv1.CronJobSpec{
					Schedule: "30 10,11,12 * * 1-5",
				},
				Status: batchv1.CronJobStatus{
					LastScheduleTime: &metav1HalfPastTheHour,
				},
			},
			now:                    *deltaTimeAfterTopOfTheHour(30*time.Hour + 30*time.Minute),
			expectedRecentTime:     deltaTimeAfterTopOfTheHour(26*time.Hour + 30*time.Minute),
			expectedEarliestTime:   *deltaTimeAfterTopOfTheHour(30 * time.Minute),
			expectedNumberOfMisses: 2,
		},
```

Reply:
Should we expect the number of misses to be five (5) here? 
- 2016-05-19T10:30 (expectedEarliestTime)
- 2016-05-19T11:30
- 2016-05-19T12:30
- 2016-05-20T10:30
- 2016-05-20T11:30
- 2016-05-20T12:30 (expectedRecentTime)

Reply:
@helayoty yes that's correct, the expectedNumberOfMisses should be 5 but since numberOfMissedSchedules is an approximate, i was concentrating more on the failure with respective expectedRecentTime

Reply:
See my other comment below about missed schedules.